### ISSUE TITLE
SIGKILL handling

### ISSUE THREAD

**Issue by bergwolf** [kata-containers/shim#4]
I'm leaving the SIGKILL handling out of the first PR, so that we can find a long term solution.

In short, we need a monitor process somewhere to see if the shim is killed, and send SIGKILL to container/process as well.

Possible candidates are:
 - shim spawns a monitor process
 - runtime monitors shim
 - higher level handling (is it possible?)

OTOH, we need to decide if we want to support shim-reconnection in some cases (like upgrade?) and how to distinguish that from shim being properly stopped.

WDYT? @laijs @sameo @sboeuf @WeiZhang555

**Comment by jodh-intel** [comment]
The runtime cannot monitor the shim since it's a transitory process. The best generic solution is for the shim to spawn a monitor process, as we do for Clear Containers 3.0.

**Comment by sboeuf** [comment]
@bergwolf take a look at https://github.com/clearcontainers/shim/tree/sboeuf/shim_Go_impl if you need an example of a monitoring process.
Basically, in this implementation in Go, I `reexec` the shim in order to spawn a child. The child will monitor the parent process relying on a pipe. If this pipe is closed without receiving the end byte, this means the parent has been killed without notice, which is a SIGKILL.

**Comment by bergwolf** [comment]
I tried containerd (and docker-containerd), and the shim process for a container is monitored by containerd and if such a shim dies, `runtime kill <container_id> 9` is called by containerd to kill the container. However, a process shim is not monitored this way.

**Comment by bergwolf** [comment]
@sboeuf I wanted to consolidate the monitor processes so that we use a single monitor for all shims for the sake of less memory footprint. Runtime can create the monitor process in two ways:
  - when called by crio/containerd via commandline
  - when called as a library

The monitor process can then scan all process.json (or state.json) files (wherever runtime saves the list of live processes for a container), match them with live processes in containers, and then send `SIGKILL` to processes that is missing in json files but alive in containers.

What do you think?

**Comment by WeiZhang555** [comment]
@bergwolf 
> * shim spawns a monitor process
> * runtime monitors shim
> * higher level handling (is it possible?)

My opinion is we leave the monitor part to high level component(option 3).

As you have discussed, containerd can handle this quite well, if cri-o doens't monitor runtime shim, then we should add codes to standard crio. When shim dies, containerd/crio could detects this and it should call something like `kata-runtime kill container SIGKILL` to kill the container

> take a look at https://github.com/clearcontainers/shim/tree/sboeuf/shim_Go_impl if you need an example of a monitoring process. 

@sboeuf 's idea looks weird, shim spawned a child then child monitor parent, what if the child process is killed? This can't fully guarantee the flow.


> OTOH, we need to decide if we want to support shim-reconnection in some cases (like upgrade?) and how to distinguish that from shim being properly stopped.

What do yo mean by "shim reconnection"? For hot-upgrade scenario, if containerd/crio is upgraded, it should reconnect to kata-shim, this part is responsibility of high level software;  but if shim dies, container or exec process dies. Shim is the smallest monitor unit which can never abort abnormally.  The live-restore part should be done at higher level.

**Comment by bergwolf** [comment]
@WeiZhang555 I was thinking abhout hot-upgrade kata-shim. The shim monitor should know the difference between a hot-upgrading shim and a shim that is killed by SIGKILL, and do not kill the hot-upgrading shim.

But since we do not share shims, it might be over-skilled to support shim hot-upgrade.

**Comment by bergwolf** [comment]
@WeiZhang555 while upper layer monitoring is preferred, the state of art is that only containerd monitors container process shims and even containerd does not monitor exec process shims. So I suggested creating a per host shim monitor in the kata runtime, which we can control.

**Comment by sboeuf** [comment]
@WeiZhang555 I don't see any reason for the child process (the one monitoring the parent) to be killed.
Anyway, I like the idea of having the higher layer (containerd or CRI-O) monitoring the shim process and send a `kata-runtime kill container SIGKILL` in case the shim is killed. My only concern is that we're actually talking about being able to rely on this upper layer that we cannot control, that's why we should have our own mechanism at first. IF we can make sure those upper layer like containerd and CRI-O will not send any SIGKILL on the shim process, but instead will use the runtime to kill the process, this would simplify things on our side too.

**Comment by WeiZhang555** [comment]
@sboeuf 

> I don't see any reason for the child process (the one monitoring the parent) to be killed.

Our test engineers will do, they kill any process they can to test the system robust, and I'm so afraid of them :-) , OOM or bad codes will also sometimes makes child process killed.

I see your points, we can do both:
* add monitor codes to cri-o if it doesn't have. (containerd is good enough)
* implement a per host shim monitor, maybe it need to be run as a sole service. I hope we can make this service optional in case upper level is doing it well in which case we don't need this shim-monitoring service.

What do you think of this?

**Comment by teawater** [comment]
@WeiZhang555 I agree with this idea.

**Comment by sboeuf** [comment]
@WeiZhang555 yes I agree with the proposal, and having one component handling such a thing could be really interesting. But this means that you expect every shim to register to this component first so that it knows which socket it should talk to in case it detects a shim being killed, right ?
Also, how do you plan for such a component to make the difference between a shim being killed and a shim being properly terminated ?
I want to make sure we don't end up with a separate component too much complex for the sole purpose of handling this case.

**Comment by WeiZhang555** [comment]
> But this means that you expect every shim to register to this component first so that it knows which socket it should talk to in case it detects a shim being killed, right ?

Yes, you're right. That's a real problem. That's why I prefer monitor from high level :)

> Also, how do you plan for such a component to make the difference between a shim being killed and a shim being properly terminated ?

One way I can think of is, let shim write a monitor pipe when it exits normally, so if it's killed the pipe would be closed abnormally.

Another problem is if per host monitor is killed, and shims are killed before monitor is pulled up by systemd, the monitor should rescan all the shim processes. So the shim process list need to be kept in disk, and when monitor is up, it need to detect liveness of processes in this list.

So I always hope high level component can do this monitor work for us, that would make thing simpler. An external monitor component makes our whole system much more complicated.

@sboeuf any more suggestions?

**Comment by teawater** [comment]
@WeiZhang555 
Per host makes the design too complex.  I think why not go back to per process monitor?

For example in https://github.com/clearcontainers/shim/tree/sboeuf/shim_Go_impl
The monitor watches the shim.

And shim can watch the monitor with "cmd.Wait" inside a goroutine.

**Comment by sboeuf** [comment]
@teawater this is what I mentioned initially in this thread, but @WeiZhang555 thinks we should consider the case where the child is killed, in this case we have nothing else left for monitoring the shim.

@WeiZhang555 I think we need to be clear about who should monitor the shim, and having two different ways of monitoring it would create some conflicts since the same process inside the VM will be signalled twice (might not be an issue, not sure).

I agree having this handled directly by the higher layer would be easier for us, but I am really wondering if this is the place it should be done. For us, because we are in the virtualized container case, and because we need a shim, we have this constraint of detecting that a shim has been SIGKILL'ed. But if you think about the runc case, there is no need for that. For a standard container, if the process receives a SIGKILL, there is no real need for any detection since the right process have been killed.

So basically what I am saying is that YES I would like to see this handled at upper layer, but I don't see any reason why higher level component would accept that.

**Comment by jshachm** [comment]
@bergwolf @sboeuf  @jodh-intel 

After working with @WeiZhang555  for debugging different strange problems around shim/proxy and qemu vm , we change a little bit of our mind about of monitoring .Not only about shim but also about all  components.

First of all, `kata-shim` of contaienr `init` process means the just container's init process in the view of  high level components. So if the `kata-shim` process didn't exit while vm is dead abnormally like `OOM` and proxy is dead also(*base on one proxy watches only one container  vm*). This will lead things we face these days that upstair will not kill shim anyway because it thinks the container is still running.This makes *Wei* and I very confused. In this case, no one  will kill shim. This will cause hang problem some time.

Is there any need for double check between proxy and shim? For the case upper stairs can't send SIGKILL to shim.Or **other good suggestions do you guys have?** 

Thx a lot~

**Comment by sboeuf** [comment]
Hi @jshachm , I am not sure I got what you're trying to say here. I cannot identify the solution you're proposing to solve the issue here. As a reminder, we're not trying to kill the shim, but we want to make sure that in case the shim receives a SIGKILL, this signal will be forwarded to the container process running inside the VM. This being due to the fact that the shim itself cannot handle such a signal. That's why we've been talking about other programs able to monitor the shim process, so that we can actually forward the SIGKILL signal from this program.

**Comment by jshachm** [comment]
@sboeuf Sry~ Maybe I should make myself more clear.

**Comment by jshachm** [comment]
@sboeuf  Sry ,maybe I should make it more clear.
I absolutely understand what you mean about the `shim`  and `SIGKILL` things. Before this time we thougt upper stair  handling would be enough. However, we have found case upper stair can't cover like I saied above. 

So we suggest besides upper stairs ,adding another way to make `shim` robust and forward `SIGKILL`  right : another light process monitoring `shim`could be needed ,and a simple way maybe **double check** between `proxy` and `shim`. Spwaning a child process maybe can work too and I will test later today to see if it works in our case. 

Thx anyway and maybe we can discuss this in slack later.

**Comment by sboeuf** [comment]
@jshachm thanks for clarifications. What you suggest about the proxy would be doable but not the best solution because the proxy could not be part of the equation in case we rely on VSOCK to communicate with the agent. In this case, there would be no need for a proxy, thus we would loose our way to monitor the shim.
That's why I was proposing the shim could fork himself and its children would monitor its parent.

**Comment by egernst** [comment]
Ah, finally reading through this (4 months later?). This is a really good issue that we may still want to discuss to make our 1.0 solution more robust.

**Comment by egernst** [comment]
@bergwolf  - any updates on this?

**Comment by raravena80** [comment]
@bergwolf any updates? 🏏

**Comment by jcvenegas** [comment]
Any update on this, some years later this is the status

shim <-proxy  -> agent
shim <  vsock > agent
shimv2 < -proxy/vsock  -> agent 

In  all cases we have a shim but shimv2 is also the runtime at the same time so all the logic could be lost.
I wonder if having the agent able to detect a connection was closed do the kill inside of the agent.

**Comment by jcvenegas** [comment]
Or as @jodh-intel suggested we can reuse the model in CC3.

**Comment by c3d** [comment]
Assigning to @bergwolf to re-evaluate if we need to keep the issue open.
